src/test/regress/upgrade/upgrade_test.py

```diff
@@ -0,0 +1,139 @@
+    for node_name in NODE_NAMES:
+        abs_data_path = os.path.abspath(os.path.join(rel_data_path, node_name))
+        command = [os.path.join(pg_path, 'pg_ctl'), '--pgdata', abs_data_path, '-U',
+                   USER, "-o '-p {}'".format(NODE_PORTS[node_name]), '--log',
+                   os.path.join(pg_path, 'logfile_' + node_name), 'start']
+        utils.run(command)
+
+
+def create_citus_extension(pg_path):
+    for port in NODE_PORTS.values():
+        utils.psql(pg_path, port, "CREATE EXTENSION citus;")
+
+
+def add_workers(pg_path):
+    for port in WORKER_PORTS:
+        command = "SELECT * from master_add_node('localhost', {port});".format(
+            port=port)
+        utils.psql(pg_path, NODE_PORTS[COORDINATOR_NAME], command)
+
+
+def run_pg_regress(pg_path, PG_SRCDIR, port, schedule):
+    command = [os.path.join(PG_SRCDIR, 'src/test/regress/pg_regress'), '--port', str(port),
+               '--schedule', schedule, '--bindir', pg_path, '--user', USER, '--dbname', DBNAME,
+               '--use-existing']
+    exit_code = utils.run(command)
+    if exit_code != 0:
+        sys.exit(exit_code)
+
+
+def citus_prepare_pg_upgrade(pg_path):
+    for port in NODE_PORTS.values():
+        utils.psql(pg_path, port, "SELECT citus_prepare_pg_upgrade();")
+
+
+def stop_databases(pg_path, rel_data_path):
+    for node_name in NODE_NAMES:
+        abs_data_path = os.path.abspath(os.path.join(rel_data_path, node_name))
+        command = [os.path.join(pg_path, 'pg_ctl'), '--pgdata', abs_data_path, '-U',
+                   USER, "-o '-p {}'".format(NODE_PORTS[node_name]), '--log',
+                   os.path.join(pg_path, 'logfile_' + node_name), 'stop']
+        utils.run(command)
+
+
+def perform_postgres_upgrade():
+    for node_name in NODE_NAMES:
+        base_new_data_path = os.path.abspath(config[NEW_DATADIR])
+        base_old_data_path = os.path.abspath(config[OLD_DATADIR])
+        with utils.cd(base_new_data_path):
+            abs_new_data_path = os.path.join(base_new_data_path, node_name)
+            abs_old_data_path = os.path.join(base_old_data_path, node_name)
+            command = [os.path.join(config[NEW_BINDIR], 'pg_upgrade'), '--username', USER,
+                       '--old-bindir', config[OLD_BINDIR], '--new-bindir', config[NEW_BINDIR],
+                       '--old-datadir', abs_old_data_path, '--new-datadir', abs_new_data_path]
+            utils.run(command)
+
+
+def citus_finish_pg_upgrade(pg_path):
+    for port in NODE_PORTS.values():
+        utils.psql(pg_path, port, "SELECT citus_finish_pg_upgrade();")
+
+
+def initialize_citus_cluster():
+    initialize_db_for_cluster(config[OLD_BINDIR], config[OLD_DATADIR])
+    start_databases(config[OLD_BINDIR], config[OLD_DATADIR])
+    create_citus_extension(config[OLD_BINDIR])
+    add_workers(config[OLD_BINDIR])
+
+
+def main():
+    initialize_temp_dir()
+    initialize_citus_cluster()
+
+    run_pg_regress(config[OLD_BINDIR], config[PG_SRCDIR],
```

Comment:
If you provide a path to pg_regress that doesn't exist there are two things that happen:
 - the test suite crashes
 - postgres keeps running

The crash we can live with, but it would be great if postgres gets shut down when our testing application crashes.

Reply:
To resolve this, we can add a try/catch/finally block as follows:
```python
    exit_code = FAIL_CODE # a nonzero code
    try:
        # here run our script
        ...
        # if we come here it means we did not get any exception and pg_regress run successfully so upgrade was successful
        exit_code = SUCCESS_CODE # zero code
    except SystemExit as code:
        # pg_regress returned nonzero code so sys.exit is called, which throws SystemExit exception.
        exit_code = code
    finally:
        # here we can stop the databases that we started and exit with our exit_code
        sys.exit(exit_code)
```

If we have a block like this, we can just check the return code of the script without having to worry about side effects. What do you think @thanodnl?

Reply:
makes sense, however this needs to be around every step in the chain:
 - initdb
 - setup citus
 - pg_regress before schedule
 - prepare_upgrade
 - initdb
 - pg_upgrade
 - pg_regress after schedule

So I wonder if python has a more generic hook that we can use on exit to clean up our databases.

Reply:
possibly something like https://docs.python.org/3/library/atexit.html